source/extensions/access_loggers/grpc/grpc_access_log_impl.h

```diff
@@ -1,40 +1,228 @@
+
+#define CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(COUNTER, GAUGE)                                   \
+  COUNTER(critical_logs_message_timeout)                                                           \
+  COUNTER(critical_logs_nack_received)                                                             \
+  COUNTER(critical_logs_ack_received)                                                              \
+  GAUGE(pending_critical_logs, Accumulate)
+
+struct GrpcCriticalAccessLogClientGrpcClientStats {
+  CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+class GrpcCriticalAccessLogClient {
+public:
+  using RequestType = envoy::service::accesslog::v3::CriticalAccessLogsMessage;
+  using ResponseType = envoy::service::accesslog::v3::CriticalAccessLogsResponse;
+
+  struct CriticalLogStream : public Grpc::AsyncStreamCallbacks<ResponseType> {
+    explicit CriticalLogStream(GrpcCriticalAccessLogClient& parent) : parent_(parent) {}
+
+    // Grpc::AsyncStreamCallbacks
+    void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
+    void onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&&) override {}
+    void onReceiveMessage(std::unique_ptr<ResponseType>&& message) override {
+      const auto& id = message->id();
+
+      switch (message->status()) {
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::ACK:
+        parent_.stats_.critical_logs_ack_received_.inc();
+        parent_.stats_.pending_critical_logs_.dec();
+        parent_.client_->onSuccess(id);
+        break;
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::NACK:
+        parent_.stats_.critical_logs_nack_received_.inc();
+        parent_.client_->onError(id);
+        break;
+      default:
+        return;
+      }
+    }
+    void onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) override {}
+    void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override {
+      parent_.client_->cleanup();
+    }
+
+    GrpcCriticalAccessLogClient& parent_;
+  };
+
+  class InflightMessageTtlManager {
+  public:
+    InflightMessageTtlManager(Event::Dispatcher& dispatcher,
+                              GrpcCriticalAccessLogClientGrpcClientStats& stats,
+                              Grpc::BufferedAsyncClient<RequestType, ResponseType>& client,
+                              std::chrono::milliseconds message_ack_timeout)
+        : dispatcher_(dispatcher), message_ack_timeout_(message_ack_timeout), stats_(stats),
+          client_(client), timer_(dispatcher_.createTimer([this] { callback(); })) {}
+
+    ~InflightMessageTtlManager() { timer_->disableTimer(); }
+
+    void setDeadline(absl::flat_hash_set<uint32_t>&& ids) {
+      auto expires_at = dispatcher_.timeSource().monotonicTime() + message_ack_timeout_;
+      deadline_.emplace(expires_at, std::move(ids));
+      timer_->enableTimer(message_ack_timeout_);
```

Comment:
what if the time is already enabled? I imagine; If the `flush` interval is shorter than `message_ack_timeout_` then this `setDeadline` potentially would be called periodically before this timer triggers, as a result this timer will never be called? (Maybe I'm worried too much though :D

Reply:
Ah, this is my mistake. This previous implementation will occur timer schedule by each time of deadline set, and unexpected behavior...